packages/ui-source-code-editor/src/SourceCodeEditor/index.tsx

```diff
@@ -0,0 +1,654 @@
+      'lineWrapping',
+      'autofocus',
+      'spellcheck',
+      'direction',
+      'dir',
+      'rtlMoveVisually',
+      'indentOnLoad',
+      'indentWithTab',
+      'indentUnit',
+      'highlightActiveLine',
+      'attachment'
+    ]
+
+    for (const prop of propsToObserve) {
+      if (!isEqual(this.props[prop], prevProps[prop])) {
+        return true
+      }
+    }
+
+    return false
+  }
+
+  get direction() {
+    // comes from the `direction` prop and
+    // falls back to the `dir` prop coming from the bidirectional decorator
+    return this.props.direction || this.props.dir
+  }
+
+  get extensions() {
+    const extensions = [
+      ...this.baseExtensions,
+
+      // our custom extensions
+      this.languageExtension,
+      this.onChangeExtension,
+      this.focusListenerExtension,
+      this.announceLineNumberExtension
+    ]
+
+    if (this.themeExtension) {
+      extensions.push(this.themeExtension)
+    }
+    if (this.props.lineNumbers) {
+      extensions.push(lineNumbers())
+    }
+    if (this.props.highlightActiveLine) {
+      extensions.push(highlightActiveLine())
+    }
+    if (this.props.highlightActiveLineGutter) {
+      extensions.push(highlightActiveLineGutter())
+    }
+    if (this.props.foldGutter) {
+      extensions.push(foldGutter())
+    }
+    if (this.props.lineWrapping) {
+      extensions.push(EditorView.lineWrapping)
+    }
+    if (this.props.editable === false) {
+      extensions.push(EditorView.editable.of(false))
+    }
+    if (this.props.readOnly) {
+      extensions.push(EditorState.readOnly.of(true))
+    }
+    if (this.props.spellcheck) {
+      extensions.push(EditorView.contentAttributes.of({ spellcheck: 'true' }))
+    }
+    if (this.direction) {
+      extensions.push(
+        EditorView.contentAttributes.of({
+          dir: this.direction
+        })
+      )
+    }
+    if (this.props.indentUnit) {
+      extensions.push(indentUnit.of(this.props.indentUnit))
+    }
+
+    return extensions
+  }
+
+  get baseExtensions() {
+    return [
+      // The extensions are based on codemirrors basic setup from 'codemirror'.
+      // It is recommended by CodeMirror, that if we want to configure
+      // our editor more precisely, we have to copy the source
+      // and adjust it as desired.
+      highlightSpecialChars(),
+      history(),
+      drawSelection(),
+      dropCursor(),
+      EditorState.allowMultipleSelections.of(true),
+      syntaxHighlighting(defaultHighlightStyle, { fallback: true }),
+      bracketMatching(),
+      closeBrackets(),
+      autocompletion(),
+      rectangularSelection(),
+      crosshairCursor(),
+      highlightSelectionMatches(),
+      indentOnInput(),
+
+      keymap.of(this.keymaps)
+    ]
+  }
+
+  get keymaps(): KeyBinding[] {
+    const keymaps: KeyBinding[] = [
+      ...closeBracketsKeymap,
+      ...this.commandKeybinding,
+      ...historyKeymap,
+      ...foldKeymap,
+      ...completionKeymap,
+      ...lintKeymap
```

Comment:
I'd link in the doc these keymaps, some users love them

Reply:
Good idea!